library/src/main/java/org/apache/apex/malhar/lib/wal/FileSystemWAL.java

```diff
@@ -0,0 +1,598 @@
+      if (tmpFilePath != null) {
+
+        Path tmpPath = new Path(tmpFilePath);
+        if (fileContext.util().exists(tmpPath)) {
+          LOG.debug("tmp path exists {}", tmpPath);
+
+          outputStream = getOutputStream(currentPointer);
+          DataInputStream inputStreamOldTmp = fileContext.open(tmpPath);
+
+          IOUtils.copyPartial(inputStreamOldTmp, currentPointer.offset, outputStream);
+
+          outputStream.flush();
+          //remove old tmp
+          inputStreamOldTmp.close();
+          LOG.debug("delete tmp {}", tmpPath);
+          fileContext.delete(tmpPath, true);
+        }
+      }
+
+      //find all valid path names
+      Set<String> validPathNames = new HashSet<>();
+      for (Map.Entry<Integer, String> entry : tmpFiles.entrySet()) {
+        if (entry.getKey() <= currentPointer.partNum) {
+          validPathNames.add(new Path(entry.getValue()).getName());
+        }
+      }
+      LOG.debug("valid names {}", validPathNames);
+
+      //there can be a failure just between the flush and the actual checkpoint which can leave some stray tmp files
+      //which aren't accounted by tmp files map
+      Path walPath = new Path(fileSystemWAL.filePath);
+      Path parentWAL = walPath.getParent();
+      if (parentWAL != null && fileContext.util().exists(parentWAL)) {
+        RemoteIterator<FileStatus> remoteIterator = fileContext.listStatus(parentWAL);
+        while (remoteIterator.hasNext()) {
+          FileStatus status = remoteIterator.next();
+          String fileName = status.getPath().getName();
+          if (fileName.startsWith(walPath.getName()) && fileName.endsWith(TMP_EXTENSION) &&
+              !validPathNames.contains(fileName)) {
+            LOG.debug("delete stray tmp {}", status.getPath());
+            fileContext.delete(status.getPath(), true);
+          }
+
+        }
+      }
+
+    }
+
+    protected void close() throws IOException
+    {
+      if (outputStream != null) {
+        flush();
+        outputStream.close();
+        outputStream = null;
+        LOG.debug("closed {}", currentPointer.partNum);
+      }
+    }
+
+    @Override
+    public int append(T entry) throws IOException
+    {
+      byte[] slice = fileSystemWAL.serde.serialize(entry);
+      int entryLength = slice.length + 4;
+
+      // rotate if needed
+      if (shouldRotate(entryLength)) {
+        rotate(true);
+      }
+
+      outputStream.writeInt(slice.length);
+      outputStream.write(slice);
+      currentPointer.offset += entryLength;
+
+      if (currentPointer.offset == fileSystemWAL.maxLength) {
+        //if the file is completed then we can rotate it. do not have to wait for next entry
+        rotate(false);
+      }
+
+      return entryLength;
+    }
+
+    protected void flush() throws IOException
+    {
+      if (outputStream != null) {
+        outputStream.flush();
+        if (outputStream instanceof FSDataOutputStream) {
+          ((FSDataOutputStream)outputStream).hflush();
```

Comment:
Please take a look at the new changes. I had to make some changes in order to read from temporary files. On local file system until the stream is closed, the data to readers isn't available.